DM-26298: Add timing information to DBB handoff manager

mxk62 opened the pull request.

mxk62 committed:
Collect metrics about files and transfer attempts

The handoff manager was not storing any data about files in the buffer
and transfer attempts it made.  As a result, it was difficult address
questions like: when a file was created, was it transferred yet, what
was the transfer rate, etc.  To address this limitation I modified the
manager to maintain a SQLite database with various pieces of information
that an operator or a monitoring system could relatively easily query to
find answers for the question mentioned earlier.

mxk62 committed:
Update documentation

Updated the 'get-started.rst' document to make it reflect recently
introduced changes.

mxk62 committed:
Update existing unit tests

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/cli.py:
```diff
@@ -0,0 +1,131 @@
+
+
+def main():
+    """Start microservices for DBB endpoint manager.
```
endpoint -> handoff?

mxk62 replied:
Handoff, of course. I must have missed it when copy-pasted the "skeleton" from the endpoint manager. Will fix.

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/defaults.py:
```diff
@@ -0,0 +1,50 @@
+    """
+
+    chunk_size = 10
+    """Number of files to processes by a transfer thread in a single session.
```
Max?  (or is it clear that sometimes number of files processed could be less than chunk_size).  Also what is a session?

mxk62 replied:
I'd say that for the new version the most accurate description of the `chunk_size` parameter would be

> a number of messages/items a manager's command retrieves from the input message queue in each iteration of its main loop (when applicable [1])

It just seemed "insanely" specific for a *general* option description.

I think I'll go with
> """Number of messages/items to retrieve from the input message queue.
>
> In the context of file transfers, it corresponds to maximal number of files that will be transferred to the endpoint site with a single transfer command ,e.g., scp or bbcp.
>"""

[1] Not every manager's command has an input message queue.

MichelleGower replied:
This comment is much better.  But still I would like to enforce that it doesn't wait until it can grab chunk_size messages. 
 Maximum number of messages/items to retrieve from the input message queue?

MichelleGower commented on doc/lsst.dbb.buffmngrs.handoff/get-started.rst:
```diff
@@ -175,27 +174,82 @@ execution, please keep in mind that:
+.. code-block:: yaml
+
+   database:
+     engine: "sqlite:///handoff.db"
```
Perhaps a more widely usable example would include a full path?

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/messages.py:
```diff
@@ -0,0 +1,91 @@
+
+@dataclass
+class TransferMsg:
+    pre_start: float = None
```
Is there a particular reason these don't exactly match the column names?  (I didn't originally notice with the times.  The just plain size and rate caught my eye.)

mxk62 replied:
Besides an aesthetic one (I tend to prefer short variable/attribute names), there's none.

I understand that database columns requires more descriptive names as you can't provide an additional comment/description (Well, at least I don't know how to do it), but `pre_start` is still pretty clear and the detailed description is right there, in class definition.

Also, you will crush my soul if you make me change `rate` to `rate_mbytes_per_sec`. :wink:

MichelleGower replied:
My concern was finding the places in the code that matched the column names (e.g., searching code for rate_mbytes_per_sec).  While I like descriptive variable names, I was not pushing for putting the units in the variable name.   If it should be easy for someone to figure out the name differences when maintaining the code, feel free to not change anything.

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/remote.py:
```diff
@@ -118,104 +121,163 @@ def run(self):
+                start = datetime.datetime.now()
+                status, stdout, stderr, dur = \
+                    execute(cmd, timeout=self.timeout)
+                total += dur
```
I must be missing something.  Why do a += here?  It is initialized a few lines above and then += here.  And it's not in a loop so isn't total just dur?

mxk62 replied:
You're right, it's not really required here. After all, the pre-transfer phase really consists of just one action.

I just implemented it this way to keep things consistent with calculating commands' duration in post-transfer phase (where there are actually two actions) and "made room" for changes in case the pre-transfer phase would ever require more than one action too.

If you think it's an overkill, I will gladly change it.

MichelleGower replied:
Please change

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/remote.py:
```diff
@@ -118,104 +121,163 @@ def run(self):
+
+                    relocated.append((batch, transfer))
+
+                # Remove batches and transfer items corresponding to failed
```
Comment a little confusing as nothing is actively being removed in the code.  Instead something like "Recreate lists without items corresponding to failed transfers".

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/remote.py:
```diff
@@ -118,104 +121,163 @@ def run(self):
-                while relocated:
-                    _, _, fn = relocated.popleft()
-                    self.done.put((head, tail, fn))
+                    relocated.append((batch, transfer))
```
Why appending batch to relocated when only going to use the transfer part in the next line (and relocated is reset at the top of this post-transfer actions?  If it is just to be consistent about structure of "relocated" throughout the code, then pick a different variable name

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/remote.py:
```diff
@@ -261,7 +323,7 @@ def run(self):
                             f"-delete")
         cmd = tpl.format(**self.params, **args)
-        status, stdout, stderr = execute(cmd, timeout=self.time)
+        status, stdout, stderr, duration = execute(cmd, timeout=self.time)
```
Added duration, but never used it.

mxk62 replied:
Yes, it is not used as I don't see it exactly as a "file-transfer-specific" command so I don't really care (maybe incorrectly) about it's duration. However, I need to catch the 4th argument which `execute()` currently returns anyway.

Are you fine with changing it to
```python
status, stdout, stderr, _ = execute(cmd, timeout=self.time)
```
or did you have something else in mind?

MichelleGower replied:
Perfect.

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/utils.py:
```diff
@@ -0,0 +1,175 @@
+        time.sleep(pause)
+
+
+def setup_dbb_conn(config):
```
Just checking whether really meant dbb or just db

mxk62 replied:
`db`, it should read `db`. Will fix.

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/messages.py:
```diff
@@ -0,0 +1,91 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+from dataclasses import dataclass
```
Module docstring?

MichelleGower commented on python/lsst/dbb/buffmngrs/handoff/utils.py:
```diff
@@ -0,0 +1,175 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+import hashlib
```
Module docstring?

MichelleGower reviewed (approved):
Some minor changes and then can merge.

mxk62 committed:
Add module-level docstrings

mxk62 committed:
Fix a typo in funtion's name

mxk62 committed:
Make selected comments more clear

mxk62 committed:
Make minor refactoring in Porter class

Following suggestions from the reviewer, I did some minor code
refactoring to make the code easier to follow.

mxk62 committed:
Pick better example database connect string

mxk62 committed:
Fix bug in Click commands

The error messages in Click commands were trying to access a
non-existent exception attribute.  Fixed it.

mxk62 committed:
Fix whitespaces here and there

mxk62 committed:
Remove unused modules and variables

mxk62 committed:
Add missing API docstrings

mxk62 committed:
Make pylint more happy with Manager class

mxk62 committed:
Simplify getting duration for pre-transfer actions
